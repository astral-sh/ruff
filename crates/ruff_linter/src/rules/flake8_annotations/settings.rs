//! Settings for the `flake-annotations` plugin.

use crate::display_settings;
use ruff_macros::CacheKey;
use std::fmt::{Display, Formatter};

#[derive(Debug, Default, CacheKey)]
#[allow(clippy::struct_excessive_bools)]
pub struct Settings {
    pub mypy_init_return: bool,
    pub suppress_dummy_args: bool,
    pub suppress_none_returning: bool,
    pub allow_star_arg_any: bool,
    pub ignore_fully_untyped: bool,
}

impl Display for Settings {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        display_settings! {
            formatter = f,
            namespace = "linter.flake8_annotations",
            fields = [
                self.mypy_init_return,
                self.suppress_dummy_args,
                self.suppress_none_returning,
                self.allow_star_arg_any,
                self.ignore_fully_untyped
            ]
        }
        Ok(())
    }
}
