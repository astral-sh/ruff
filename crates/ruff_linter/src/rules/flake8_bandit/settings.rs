//! Settings for the `flake8-bandit` plugin.

use ruff_macros::CacheKey;

pub fn default_tmp_dirs() -> Vec<String> {
    ["/tmp", "/var/tmp", "/dev/shm"]
        .map(ToString::to_string)
        .to_vec()
}

#[derive(Debug, CacheKey)]
pub struct Settings {
    pub hardcoded_tmp_directory: Vec<String>,
    pub check_typed_exception: bool,
}

impl Default for Settings {
    fn default() -> Self {
        Self {
            hardcoded_tmp_directory: default_tmp_dirs(),
            check_typed_exception: false,
        }
    }
}
