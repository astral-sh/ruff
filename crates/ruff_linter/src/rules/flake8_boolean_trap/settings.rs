//! Settings for the `flake8-boolean-trap` plugin.

use std::fmt;

use ruff_macros::CacheKey;

use crate::display_settings;

#[derive(Debug, Clone, CacheKey, Default)]
pub struct Settings {
    pub extend_allowed_calls: Vec<String>,
}

impl fmt::Display for Settings {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        display_settings! {
            formatter = f,
            namespace = "linter.flake8_boolean_trap",
            fields = [
                self.extend_allowed_calls | array,
            ]
        }
        Ok(())
    }
}
