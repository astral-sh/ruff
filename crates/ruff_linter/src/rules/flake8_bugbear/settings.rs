//! Settings for the `flake8-bugbear` plugin.

use ruff_macros::CacheKey;

#[derive(Debug, Default, CacheKey)]
pub struct Settings {
    pub extend_immutable_calls: Vec<String>,
}
