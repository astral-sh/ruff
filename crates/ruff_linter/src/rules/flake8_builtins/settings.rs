//! Settings for the `flake8-builtins` plugin.

use crate::display_settings;
use ruff_macros::CacheKey;
use std::fmt::{Display, Formatter};

#[derive(Debug, Default, CacheKey)]
pub struct Settings {
    pub builtins_ignorelist: Vec<String>,
}

impl Display for Settings {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        display_settings! {
            formatter = f,
            namespace = "linter.flake8_builtins",
            fields = [
                self.builtins_ignorelist | array
            ]
        }
        Ok(())
    }
}
