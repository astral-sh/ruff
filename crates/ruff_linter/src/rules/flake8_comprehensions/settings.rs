//! Settings for the `flake8-comprehensions` plugin.

use ruff_macros::CacheKey;

#[derive(Debug, Default, CacheKey)]
pub struct Settings {
    pub allow_dict_calls_with_keyword_arguments: bool,
}
