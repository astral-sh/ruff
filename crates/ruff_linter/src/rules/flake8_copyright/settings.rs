//! Settings for the `flake8-copyright` plugin.

use once_cell::sync::Lazy;
use regex::Regex;
use std::fmt::{Display, Formatter};

use crate::display_settings;
use ruff_macros::CacheKey;

#[derive(Debug, CacheKey)]
pub struct Settings {
    pub notice_rgx: Regex,
    pub author: Option<String>,
    pub min_file_size: usize,
}

pub static COPYRIGHT: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?i)Copyright\s+(\(C\)\s+)?\d{4}(-\d{4})*").unwrap());

impl Default for Settings {
    fn default() -> Self {
        Self {
            notice_rgx: COPYRIGHT.clone(),
            author: None,
            min_file_size: 0,
        }
    }
}

impl Display for Settings {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        display_settings! {
            formatter = f,
            namespace = "linter.flake8_copyright",
            fields = [
                self.notice_rgx,
                self.author | optional,
                self.min_file_size,
            ]
        }
        Ok(())
    }
}
