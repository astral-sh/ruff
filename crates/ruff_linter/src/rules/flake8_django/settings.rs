//! Settings for the `flake8-django` plugin.

use crate::display_settings;
use ruff_macros::CacheKey;
use std::fmt::{Display, Formatter};

#[derive(Debug, Clone, CacheKey)]
pub struct Settings {
    pub additional_path_functions: Vec<String>,
}

impl Default for Settings {
    fn default() -> Self {
        Self {
            additional_path_functions: vec![],
        }
    }
}

impl Display for Settings {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        display_settings! {
            formatter = f,
            namespace = "linter.flake8_django",
            fields = [
                self.additional_path_functions | array
            ]
        }
        Ok(())
    }
}
