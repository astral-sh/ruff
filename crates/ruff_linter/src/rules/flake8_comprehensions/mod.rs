//! Rules from [flake8-comprehensions](https://pypi.org/project/flake8-comprehensions/).
mod fixes;
pub(crate) mod rules;
pub mod settings;

#[cfg(test)]
mod tests {
    use std::path::Path;

    use anyhow::Result;
    use test_case::test_case;

    use crate::assert_messages;
    use crate::registry::Rule;
    use crate::settings::LinterSettings;
    use crate::test::test_path;

    #[test_case(Rule::UnnecessaryCallAroundSorted, Path::new("C413.py"))]
    #[test_case(Rule::UnnecessaryCollectionCall, Path::new("C408.py"))]
    #[test_case(Rule::UnnecessaryComprehension, Path::new("C416.py"))]
    #[test_case(Rule::UnnecessaryComprehensionAnyAll, Path::new("C419.py"))]
    #[test_case(Rule::UnnecessaryDoubleCastOrProcess, Path::new("C414.py"))]
    #[test_case(Rule::UnnecessaryGeneratorDict, Path::new("C402.py"))]
    #[test_case(Rule::UnnecessaryGeneratorList, Path::new("C400.py"))]
    #[test_case(Rule::UnnecessaryGeneratorSet, Path::new("C401.py"))]
    #[test_case(Rule::UnnecessaryListCall, Path::new("C411.py"))]
    #[test_case(Rule::UnnecessaryListComprehensionDict, Path::new("C404.py"))]
    #[test_case(Rule::UnnecessaryListComprehensionSet, Path::new("C403.py"))]
    #[test_case(Rule::UnnecessaryLiteralDict, Path::new("C406.py"))]
    #[test_case(Rule::UnnecessaryLiteralSet, Path::new("C405.py"))]
    #[test_case(Rule::UnnecessaryLiteralWithinDictCall, Path::new("C418.py"))]
    #[test_case(Rule::UnnecessaryLiteralWithinListCall, Path::new("C410.py"))]
    #[test_case(Rule::UnnecessaryLiteralWithinTupleCall, Path::new("C409.py"))]
    #[test_case(Rule::UnnecessaryMap, Path::new("C417.py"))]
    #[test_case(Rule::UnnecessarySubscriptReversal, Path::new("C415.py"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.noqa_code(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("flake8_comprehensions").join(path).as_path(),
            &LinterSettings::for_rule(rule_code),
        )?;
        assert_messages!(snapshot, diagnostics);
        Ok(())
    }

    #[test_case(Rule::UnnecessaryCollectionCall, Path::new("C408.py"))]
    fn allow_dict_calls_with_keyword_arguments(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!(
            "{}_{}_allow_dict_calls_with_keyword_arguments",
            rule_code.noqa_code(),
            path.to_string_lossy()
        );
        let diagnostics = test_path(
            Path::new("flake8_comprehensions").join(path).as_path(),
            &LinterSettings {
                flake8_comprehensions: super::settings::Settings {
                    allow_dict_calls_with_keyword_arguments: true,
                },
                ..LinterSettings::for_rule(rule_code)
            },
        )?;
        assert_messages!(snapshot, diagnostics);
        Ok(())
    }
}
