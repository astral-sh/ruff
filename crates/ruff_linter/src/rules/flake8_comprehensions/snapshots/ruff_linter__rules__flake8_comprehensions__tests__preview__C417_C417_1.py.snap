---
source: crates/ruff_linter/src/rules/flake8_comprehensions/mod.rs
---
C417_1.py:7:10: C417 [*] Unnecessary `map()` usage (rewrite using a generator expression)
  |
5 | def overshadowed_list():
6 |     list = ...
7 |     list(map(lambda x: x, []))
  |          ^^^^^^^^^^^^^^^^^^^^ C417
  |
  = help: Replace `map()` with a generator expression

ℹ Unsafe fix
4 4 | 
5 5 | def overshadowed_list():
6 6 |     list = ...
7   |-    list(map(lambda x: x, []))
  7 |+    list((x for x in []))
8 8 | 
9 9 | 
10 10 | set(map(lambda x, y: x, nums, nums))

C417_1.py:10:1: C417 [*] Unnecessary `map()` usage (rewrite using a set comprehension)
   |
10 | set(map(lambda x, y: x, nums, nums))
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ C417
11 |
12 | list(map(lambda x: (a := x), foo))
   |
   = help: Replace `map()` with a set comprehension

ℹ Unsafe fix
7  7  |     list(map(lambda x: x, []))
8  8  | 
9  9  | 
10    |-set(map(lambda x, y: x, nums, nums))
   10 |+{x for x, y in zip(nums, nums)}
11 11 | 
12 12 | list(map(lambda x: (a := x), foo))
13 13 | 

C417_1.py:12:1: C417 [*] Unnecessary `map()` usage (rewrite using a list comprehension)
   |
10 | set(map(lambda x, y: x, nums, nums))
11 |
12 | list(map(lambda x: (a := x), foo))
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ C417
13 |
14 | list(map(lambda x: (a for a in \
   |
   = help: Replace `map()` with a list comprehension

ℹ Unsafe fix
9  9  | 
10 10 | set(map(lambda x, y: x, nums, nums))
11 11 | 
12    |-list(map(lambda x: (a := x), foo))
   12 |+[(a := x) for x in foo]
13 13 | 
14 14 | list(map(lambda x: (a for a in \
15 15 |                     range(x)), foo))

C417_1.py:14:1: C417 [*] Unnecessary `map()` usage (rewrite using a list comprehension)
   |
12 |   list(map(lambda x: (a := x), foo))
13 |
14 | / list(map(lambda x: (a for a in \
15 | |                     range(x)), foo))
   | |____________________________________^ C417
16 |
17 |   dict(map(lambda k, v: (k, v), keys, values))
   |
   = help: Replace `map()` with a list comprehension

ℹ Unsafe fix
11 11 | 
12 12 | list(map(lambda x: (a := x), foo))
13 13 | 
14    |-list(map(lambda x: (a for a in \
15    |-                    range(x)), foo))
   14 |+[(a for a in \
   15 |+                    range(x)) for x in foo]
16 16 | 
17 17 | dict(map(lambda k, v: (k, v), keys, values))
18 18 | 

C417_1.py:17:1: C417 [*] Unnecessary `map()` usage (rewrite using a dict comprehension)
   |
15 |                     range(x)), foo))
16 |
17 | dict(map(lambda k, v: (k, v), keys, values))
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ C417
18 |
19 | dict(
   |
   = help: Replace `map()` with a dict comprehension

ℹ Unsafe fix
14 14 | list(map(lambda x: (a for a in \
15 15 |                     range(x)), foo))
16 16 | 
17    |-dict(map(lambda k, v: (k, v), keys, values))
   17 |+{k: v for k, v in zip(keys, values)}
18 18 | 
19 19 | dict(
20 20 |     map(

C417_1.py:19:1: C417 [*] Unnecessary `map()` usage (rewrite using a dict comprehension)
   |
17 |   dict(map(lambda k, v: (k, v), keys, values))
18 |
19 | / dict(
20 | |     map(
21 | |         lambda k, v: (
22 | |             (  # Foo
23 | |                 k
24 | |             ),
25 | |             v **2
26 | |         ),
27 | |         keys, values
28 | |     )
29 | | )
   | |_^ C417
30 |
31 |   list(map(lambda x: list[...], foo))
   |
   = help: Replace `map()` with a dict comprehension

ℹ Unsafe fix
16 16 | 
17 17 | dict(map(lambda k, v: (k, v), keys, values))
18 18 | 
19    |-dict(
20    |-    map(
21    |-        lambda k, v: (
22    |-            (  # Foo
   19 |+{(  # Foo
23 20 |                 k
24    |-            ),
25    |-            v **2
26    |-        ),
27    |-        keys, values
28    |-    )
29    |-)
   21 |+            ): v **2 for k, v in zip(keys, values)}
30 22 | 
31 23 | list(map(lambda x: list[...], foo))
32 24 | 

C417_1.py:31:1: C417 [*] Unnecessary `map()` usage (rewrite using a list comprehension)
   |
29 | )
30 |
31 | list(map(lambda x: list[...], foo))
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ C417
   |
   = help: Replace `map()` with a list comprehension

ℹ Unsafe fix
28 28 |     )
29 29 | )
30 30 | 
31    |-list(map(lambda x: list[...], foo))
   31 |+[list[...] for x in foo]
32 32 | 
33 33 | 
34 34 | def unfixable():

C417_1.py:36:5: C417 Unnecessary `map()` usage (rewrite using a generator expression)
   |
34 | def unfixable():
35 |     zip = []
36 |     map(lambda x, y: x + y + 1, a, b)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ C417
   |
   = help: Replace `map()` with a generator expression

C417_1.py:39:1: C417 [*] Unnecessary `map()` usage (rewrite using a dict comprehension)
   |
39 | / dict(
40 | |     map(
41 | |         lambda x, y: (
42 | |             a := 0,
43 | |             b := 1
44 | |         ),
45 | |         foo, bar
46 | |     )
47 | | )
   | |_^ C417
   |
   = help: Replace `map()` with a dict comprehension

ℹ Unsafe fix
36 36 |     map(lambda x, y: x + y + 1, a, b)
37 37 | 
38 38 | 
39    |-dict(
40    |-    map(
41    |-        lambda x, y: (
42    |-            a := 0,
43    |-            b := 1
44    |-        ),
45    |-        foo, bar
46    |-    )
47    |-)
   39 |+{(a := 0): (b := 1) for x, y in zip(foo, bar)}
48 40 | 
49 41 | 
50 42 | dict(

C417_1.py:50:1: C417 [*] Unnecessary `map()` usage (rewrite using a dict comprehension)
   |
50 | / dict(
51 | |     map(
52 | |         lambda x, y: (
53 | |             list[...],
54 | |         # Comment
55 | |             (  # Comment
56 | |                 a for a \
57 | |                 in b
58 | |             )
59 | |         ),
60 | |         foo, bar
61 | |     )
62 | | )
   | |_^ C417
   |
   = help: Replace `map()` with a dict comprehension

ℹ Unsafe fix
47 47 | )
48 48 | 
49 49 | 
50    |-dict(
51    |-    map(
52    |-        lambda x, y: (
53    |-            list[...],
54    |-        # Comment
55    |-            (  # Comment
   50 |+{list[...]: (  # Comment
56 51 |                 a for a \
57 52 |                 in b
58    |-            )
59    |-        ),
60    |-        foo, bar
61    |-    )
62    |-)
   53 |+            ) for x, y in zip(foo, bar)}
63 54 | 
64 55 | 
65 56 | ### No errors
